File: lib/track.rb
class Spotbot::Track
  attr_reader :uri

  def initialize(uri)
    @uri = uri
  end

  def as_json
    {
      title: track.name,
      artists: track.artists.map(&:name),
      duration: track.duration_ms,
      image: image,
      uri: uri,
    }
  end

  private

  def image
    return unless track.album && track.album.images && track.album.images.first
    track.album.images.first["url"]
  end

  def track
    @track ||= RSpotify::Track.find uri_id
  end

  def uri_id
    uri.gsub("spotify:track:", "")
  end
end

File: lib/track.js
"use strict";
module.exports = function(spotify){
  function Track(uriOrObject) {
    if (typeof(uriOrObject) === "string") {
      this.uri = uriOrObject;
    } else {
      this.uri = uriOrObject.link;
      this.link = uriOrObject;
    }
  }

  Track.prototype.getLink = function() {
    if (this.link) {
      return this.link;
    } else {
      this.link = spotify.createFromLink(this.uri);
      return this.link;
    }
  };

  Track.prototype.json = function() {
    return {
      uri: this.uri,
      artists: this.artists(),
      title: this.title(),
      duration: this.duration(),
      image: this.image()
    };
  };

  Track.prototype.artists = function() {
    var track = this.getLink();
    if (track) {
      return track.artists.map(function(artist){
        return artist.name;
      });
    }
  };

  Track.prototype.title = function() {
    var track = this.getLink();
    if (track) {
      return track.name;
    }
  };

  Track.prototype.duration = function() {
    var track = this.getLink();
    if (track) {
      return track.duration * 1000;
    }
  };

  Track.prototype.image = function() {
    return "";
  };


  return Track;
};
